pkg/operator/awsTagController.go

```diff
@@ -0,0 +1,324 @@
+		c.queue.Forget(obj)
+		klog.V(5).Infof("AWSController: event from workqueue successfully processed")
+	}
+	return true
+}
+
+// sync method is defined for handling the operations required
+// on receiving a informer event.
+// Fetches image registry config data, required for obtaining
+// the S3 bucket configuration and creating a driver out of it
+func (c *AWSTagController) sync() error {
+	return c.syncTags()
+}
+
+// syncTags fetches user tags from Infrastructure resource, which
+// is then compared with the tags configured for the created S3 bucket
+// fetched using the driver object passed and updates if any new tags.
+func (c *AWSTagController) syncTags() error {
+	cr, err := c.imageRegistryConfigClient.Get(
+		context.Background(),
+		defaults.ImageRegistryResourceName,
+		metav1.GetOptions{},
+	)
+	if err != nil {
+		return err
+	}
+
+	// if s3 storage config is missing, must be
+	// non-AWS platform, so not treating it as error
+	if cr.Spec.Storage.S3 == nil {
+		return nil
+	}
+
+	// make a copy to avoid changing the cached data
+	cr = cr.DeepCopy()
+
+	infra, err := c.infraConfigClient.Get(
+		context.Background(),
+		defaults.InfrastructureResourceName,
+		metav1.GetOptions{},
+	)
+	if err != nil {
+		klog.Errorf("failed to fetch Infrastructure resource: %v", err)
+		return err
+	}
+
+	// tags deletion is not supported. Should the user remove it from
+	// PlatformStatus will be looked up for retaining the tag
+	infraTagSet := make(map[string]string)
+	mergePlatformStatusTags(infra, infraTagSet)
+	klog.V(5).Infof("tags read from Infrastructure resource: %v", infraTagSet)
+
+	// Create a driver with the current configuration
+	ctx := context.Background()
+	driver := s3.NewDriver(ctx, cr.Spec.Storage.S3, &c.listers.StorageListers, c.featureGateAccessor)
+
+	s3TagSet, err := driver.GetStorageTags()
+	if err != nil {
+		klog.Errorf("failed to fetch storage tags: %v", err)
+		return err
+	}
+	klog.V(5).Infof("tags read from storage resource: %v", s3TagSet)
+
+	tagUpdatedCount := compareS3InfraTagSet(s3TagSet, infraTagSet)
+	if tagUpdatedCount > 0 {
+		if err := driver.PutStorageTags(s3TagSet); err != nil {
+			klog.Errorf("failed to update/delete tagset of %s s3 bucket: %v", driver.ID(), err)
+			c.event.Warningf("UpdateAWSTags",
+				"Failed to update/delete tagset of %s s3 bucket", driver.ID())
+		}
+		klog.Infof("successfully updated/deleted %d tags, tagset: %+v", tagUpdatedCount, s3TagSet)
+		c.event.Eventf("UpdateAWSTags",
+			"Successfully updated/deleted tagset of %s s3 bucket", driver.ID())
+	}
+
+	return nil
+}
+
+// mergePlatformStatusTags is for reading and merging user tags present in both
+// Platform Status of Infrastructure config.
+// There could be scenarios(upgrade, user deletes) where user tags could be missing
+// from the Platform Status, hence using Status too to avoid said scenarios.
+// If a tag exists in Status.
+func mergePlatformStatusTags(infra *configv1.Infrastructure, infraTagSet map[string]string) {
```

Comment:
Is there a specific reason why this function receives `infraTagSet` as a parameter, instead of simply returning it? 
The caller (on line 239) declares `infraTagSet` before calling `mergePlatformStatusTags`, i.e:
```
infraTagSet := make(map[string]string)
mergePlatformStatusTags(infra, infraTagSet)
```
So I'm inclined to think there is no reason to receive this as a parameter, but instead declare `infraTagSet` inside `mergePlatformStatusTags` and simply return it, i.e:
```suggestion
func mergePlatformStatusTags(infra *configv1.Infrastructure) map[string]string {
	infraTagSet := make(map[string]string)
```
Then in the end of the function just `return infraTagSet`.

There's also room for improvement in the function name. This function seem to be more about filtering out invalid tags from the platform status... How about naming it something like `filterPlatformStatusTags` instead?

WDYT?

Reply:
Agreed!!